File: src/archview/extract/typescript.py
"""Turn what the TypeScript compiler reports into the model (ADR 0010).

`typescript.mjs` runs in Node with the analysed repo's own `typescript` package and
prints raw facts per file (see `read_facts`); every decision about ids, kinds,
externals and warnings is made here. Ids are paths below the source root, prefixed by
the project and keeping the file extension, split by '/'.
"""

from __future__ import annotations

import re
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

from archview.model.graph import ExtractionWarning, Import, Model, Node
from archview.model.names import ancestors, parent

SEP = "/"
CODE = re.compile(r"\.(ts|tsx|mts|cts|js|jsx|mjs|cjs)$")
DECLARATION = re.compile(r"\.d\.(ts|mts|cts)$")
NON_CODE = re.compile(
    r"\.(png|jpe?g|gif|svg|webp|avif|ico|bmp|css|scss|sass|less|json|md|mdx|html|txt"
    r"|woff2?|ttf|otf|eot|mp3|mp4|wav|webm|lottie|riv|wasm|graphql|gql|ya?ml)$",
    re.IGNORECASE,
)
ROOT_CONFIG = re.compile(r"^[^/]*\.config\.[^/]+$")  # vite.config.ts, next.config.mjs


@dataclass(frozen=True, slots=True)
class Target:
    """Where one import goes: `internal` (a repo file), `external` (an npm package),
    `drop` (builtin, declaration, asset), `unresolved` or `dynamic` (warnings)."""

    kind: str
    name: str | None = None


def _package_name(specifier: str) -> str:
    """`react-native/Libraries/x` -> `react-native`; `@expo/vector-icons/x` -> `@expo/vector-icons`."""  # noqa: E501
    parts = specifier.split("/")
    return "/".join(parts[:2]) if specifier.startswith("@") and len(parts) > 1 else parts[0]


def _is_bare(specifier: str) -> bool:
    return not specifier.startswith((".", "/"))


def _external_name(specifier: str, resolved: str) -> str:
    """The package a bare specifier names; else the package or folder the file sits in."""
    if _is_bare(specifier) and not specifier.startswith("~"):
        return _package_name(specifier)
    parts = resolved.split("/")
    if "node_modules" in parts:
        last = max(i for i, p in enumerate(parts) if p == "node_modules")
        return _package_name("/".join(parts[last + 1 :]))
    ups = next((i for i, p in enumerate(parts) if p != ".."), len(parts) - 1)
    return SEP.join(parts[: ups + 1])


def classify(fact: dict[str, Any]) -> Target:
    specifier, resolved = fact["specifier"], fact["resolved"]
    if fact["dynamic"]:
        return Target("dynamic")
    if fact["builtin"]:
        return Target("drop")
    if resolved is not None:
        return _resolved_target(specifier, resolved)
    if NON_CODE.search(specifier.split("?")[0]):
        return Target("drop")
    if not _is_bare(specifier) or fact["alias"]:
        return Target("unresolved")
    return Target("external", _package_name(specifier))


def _resolved_target(specifier: str, resolved: str) -> Target:
    if "node_modules" in resolved.split("/") or resolved.startswith("../"):
        return Target("external", _external_name(specifier, resolved))
    if DECLARATION.search(resolved) or not CODE.search(resolved):
        return Target("drop")
    return Target("internal", resolved)


def find_source_root(files: Iterable[str]) -> str:
    """The one top-level directory every file lies under (`src`), else '' (the repo)."""
    tops = {f.split(SEP)[0] if SEP in f else "" for f in files}
    return next(iter(tops)) if len(tops) == 1 and "" not in tops else ""


def _below(file: str, root: str) -> bool:
    return not root or file.startswith(root + SEP)


def build_model(facts: dict[str, Any], project: str, source_root: str | None = None) -> Model:
    """The model of one TypeScript project; `source_root` (relative to the repo) is
    found with `find_source_root` when not given, and files outside it are left out."""
    files = {f["file"]: f for f in facts["files"] if not ROOT_CONFIG.match(f["file"])}
    targets = {
        (file, index): classify(fact)
        for file, entry in files.items()
        for index, fact in enumerate(entry["imports"])
    }
    internal = set(files) | {
        t.name
        for t in targets.values()
        if t.kind == "internal" and t.name and not ROOT_CONFIG.match(t.name)
    }
    root = find_source_root(internal) if source_root is None else source_root.strip(SEP)
    ids = {
        file: f"{project}{SEP}{file[len(root) + 1 :] if root else file}"
        for file in sorted(internal)
        if _below(file, root)
    }
    imports, warnings = _imports(files, targets, ids, project)
    externals = sorted({i.imported for i in imports} - set(ids.values()))
    return Model(
        project=project,
        nodes=(*_tree(project, ids, files), *(Node(e, None, "external") for e in externals)),
        imports=tuple(sorted(imports, key=lambda i: (i.importer, i.imported, i.line))),
        language="typescript",
        warnings=tuple(sorted(warnings, key=lambda w: (w.module, w.line))),
        separator=SEP,
    )


def _tree(project: str, ids: dict[str, str], files: dict[str, Any]) -> list[Node]:
    """The project, every directory on the way to a file, and the files."""
    nodes = {project: Node(project, None, "package")}
    for file, node_id in ids.items():
        for directory in ancestors(node_id, SEP)[1:-1]:
            nodes.setdefault(directory, Node(directory, parent(directory, SEP), "package"))
        abstract = bool(files.get(file, {}).get("abstract"))
        nodes[node_id] = Node(node_id, parent(node_id, SEP), "module", file, abstract)
    return sorted(nodes.values(), key=lambda n: n.id)


def _imports(files, targets, ids, project) -> tuple[list[Import], list[ExtractionWarning]]:
    imports: list[Import] = []
    warnings: list[ExtractionWarning] = []
    for file in sorted(f for f in files if f in ids):
        importer = ids[file]
        for index, fact in enumerate(files[file]["imports"]):
            target = targets[(file, index)]
            if target.kind in ("dynamic", "unresolved"):
                kind = "dynamic_import" if target.kind == "dynamic" else "unresolved_import"
                warnings.append(
                    ExtractionWarning(
                        kind, importer, file, fact["line"], fact["text"], fact["specifier"]
                    )
                )
                continue
            imported = _imported(target, ids, project)
            if imported is None or imported == importer:
                continue
            imports.append(
                Import(
                    importer,
                    imported,
                    file,
                    fact["line"],
                    fact["text"],
                    fact["type_only"],
                    fact["lazy"],
                )
            )
    return imports, warnings


def _imported(target: Target, ids: dict[str, str], project: str) -> str | None:
    if target.kind == "internal":
        return ids.get(target.name or "")
    if target.kind == "external" and target.name and target.name != project:
        return target.name  # a package importing itself by name would clash with the root
    return None

File: src/archview/extract/typescript.mjs
// archview: what the TypeScript compiler knows about a project's imports (ADR 0010).
//
//   node typescript.mjs <repo> <tsconfig>
//
// Loads the analysed repo's own `typescript` package, reads the tsconfig (following
// `extends` and `references`) and prints {"files": [...]} as JSON; the format is
// documented on `read_facts` in typescript.py, which builds the model from it. The
// source is parsed, never executed. Any problem is one line on stderr and exit code 2.

import fs from "node:fs";
import path from "node:path";
import { builtinModules, createRequire } from "node:module";

const CODE = /\.(ts|tsx|mts|cts|js|jsx|mjs|cjs)$/;
const DECLARATION = /\.d\.(ts|mts|cts)$/;
const NO_INPUTS = 18003; // a solution-style tsconfig lists no files of its own
const STATEMENT = 200; // the evidence line in `check`, `why` and the viewer; never a whole file

function fail(message) {
  process.stderr.write(`${message}\n`);
  process.exit(2);
}

// An unreadable file or a throw out of the compiler is a failure like any other: one
// line, exit 2 - never a stack trace for read_facts to hand to the user.
process.on("uncaughtException", (error) => fail(error.message));

const [repoArg, tsconfigArg] = process.argv.slice(2);
if (!repoArg || !tsconfigArg) fail("usage: node typescript.mjs <repo> <tsconfig>");
if (!fs.existsSync(repoArg)) fail(`no such directory: ${repoArg}`);
const repo = fs.realpathSync(path.resolve(repoArg));
const rel = (file) => path.relative(repo, file).split(path.sep).join("/");
const real = (file) => {
  try {
    return fs.realpathSync(file);
  } catch {
    return file;
  }
};

function loadTypeScript() {
  try {
    return createRequire(path.join(repo, "package.json"))("typescript");
  } catch {
    return fail(`typescript not found in ${path.join(repo, "node_modules")}; run npm install`);
  }
}

const ts = loadTypeScript();
const message = (diagnostic) => ts.flattenDiagnosticMessageText(diagnostic.messageText, " ");

function parseConfig(configPath) {
  const host = {
    ...ts.sys,
    onUnRecoverableConfigFileDiagnostic: (d) => fail(`${rel(configPath)}: ${message(d)}`),
  };
  const parsed = ts.getParsedCommandLineOfConfigFile(configPath, {}, host);
  const errors = parsed.errors.filter(
    (d) => d.category === ts.DiagnosticCategory.Error && d.code !== NO_INPUTS,
  );
  if (errors.length > 0) fail(`${rel(configPath)}: ${message(errors[0])}`);
  return parsed;
}

function projects(configPath, found = new Map()) {
  if (found.has(configPath)) return found;
  const parsed = parseConfig(configPath);
  found.set(configPath, parsed);
  for (const reference of parsed.projectReferences ?? []) {
    projects(real(ts.resolveProjectReferencePath(reference)), found);
  }
  return found;
}

function aliasInsideRepo(specifier, options) {
  const base = options.baseUrl ?? options.pathsBasePath ?? repo;
  for (const [pattern, targets] of Object.entries(options.paths ?? {})) {
    const star = pattern.indexOf("*");
    const prefix = star < 0 ? pattern : pattern.slice(0, star);
    const suffix = star < 0 ? "" : pattern.slice(star + 1);
    const matched =
      star < 0
        ? specifier === pattern
        : specifier.length >= prefix.length + suffix.length &&
          specifier.startsWith(prefix) &&
          specifier.endsWith(suffix);
    if (!matched) continue;
    const middle = star < 0 ? "" : specifier.slice(prefix.length, specifier.length - suffix.length);
    return targets.some((target) => {
      const where = rel(path.resolve(base, target.replace("*", middle)));
      return !where.startsWith("..") && !where.split("/").includes("node_modules");
    });
  }
  return false;
}

function resolver(options) {
  const cache = ts.createModuleResolutionCache(repo, (name) => name, options);
  return (specifier, containingFile, mode) => {
    const found = ts.resolveModuleName(
      specifier, containingFile, options, ts.sys, cache, undefined, mode,
    ).resolvedModule;
    return found ? rel(real(found.resolvedFileName)) : null;
  };
}

const hasModifier = (node, kind) => (node.modifiers ?? []).some((m) => m.kind === kind);

function isImportCall(node) {
  return (
    ts.isCallExpression(node) &&
    (node.expression.kind === ts.SyntaxKind.ImportKeyword ||
      (ts.isIdentifier(node.expression) && node.expression.text === "require"))
  );
}

function allTypeOnly(clause) {
  if (!clause) return false;
  if (clause.isTypeOnly) return true;
  const named = clause.namedBindings;
  return (
    !clause.name &&
    !!named &&
    ts.isNamedImports(named) &&
    named.elements.length > 0 &&
    named.elements.every((element) => element.isTypeOnly)
  );
}

// The whole statement on one line: a wrapped `import {\n  one,\n  two,\n} from "./b";`
// reads as `import { one, two } from "./b";`, truncated so one pathological statement
// cannot bloat the JSON.
function statementText(node, source) {
  const text = source.text.slice(node.getStart(source), node.end).replace(/\s+/g, " ").trim();
  return text.length > STATEMENT ? `${text.slice(0, STATEMENT - 1)}…` : text;
}

function scan(file, options, resolve) {
  const text = fs.readFileSync(file, "utf8");
  const source = ts.createSourceFile(file, text, ts.ScriptTarget.Latest, true);
  const imports = [];
  const exported = { types: false, values: false };
  let abstractClass = false;

  const add = (node, literal, typeOnly, lazy) => {
    const line = source.getLineAndCharacterOfPosition(node.getStart(source)).line + 1;
    const common = { line, text: statementText(node, source), type_only: typeOnly, lazy };
    if (!literal) {
      imports.push({ specifier: null, resolved: null, ...common, dynamic: true, builtin: false, alias: false });
      return;
    }
    const specifier = literal.text;
    const mode = ts.getModeForUsageLocation?.(source, literal, options);
    const resolved = resolve(specifier, file, mode);
    const builtin =
      specifier.startsWith("node:") || (!resolved && builtinModules.includes(specifier.split("/")[0]));
    const alias = !resolved && aliasInsideRepo(specifier, options);
    imports.push({ specifier, resolved, ...common, dynamic: false, builtin, alias });
  };

  const visit = (node, inFunction) => {
    if (ts.isImportDeclaration(node) && ts.isStringLiteral(node.moduleSpecifier)) {
      add(node, node.moduleSpecifier, allTypeOnly(node.importClause), false);
    } else if (ts.isExportDeclaration(node)) {
      if (node.moduleSpecifier && ts.isStringLiteral(node.moduleSpecifier)) {
        add(node, node.moduleSpecifier, node.isTypeOnly, false);
      }
      exported[node.isTypeOnly ? "types" : "values"] = true;
    } else if (ts.isImportEqualsDeclaration(node) && ts.isExternalModuleReference(node.moduleReference)) {
      const expression = node.moduleReference.expression;
      add(node, ts.isStringLiteral(expression) ? expression : null, node.isTypeOnly, inFunction);
    } else if (isImportCall(node)) {
      const [argument] = node.arguments;
      const literal = argument && ts.isStringLiteralLike(argument) ? argument : null;
      add(node, literal, false, node.expression.kind === ts.SyntaxKind.ImportKeyword || inFunction);
    } else if (ts.isExportAssignment(node)) {
      exported.values = true;
    }
    if (ts.isClassDeclaration(node) && hasModifier(node, ts.SyntaxKind.AbstractKeyword)) {
      abstractClass = true;
    }
    if (node.parent === source && hasModifier(node, ts.SyntaxKind.ExportKeyword)) {
      const typeOnly = ts.isInterfaceDeclaration(node) || ts.isTypeAliasDeclaration(node);
      exported[typeOnly ? "types" : "values"] = true;
    }
    const nested = inFunction || ts.isFunctionLike(node);
    ts.forEachChild(node, (child) => visit(child, nested));
  };
  ts.forEachChild(source, (child) => visit(child, false));

  return { file: rel(file), abstract: abstractClass || (exported.types && !exported.values), imports };
}

const configPath = path.resolve(tsconfigArg);
if (!fs.existsSync(configPath)) fail(`no ${rel(configPath)} in ${repo}`);

const owners = new Map(); // file -> the compiler options of the first tsconfig listing it
for (const parsed of projects(real(configPath)).values()) {
  for (const name of parsed.fileNames) {
    const file = real(name);
    // A `references` entry can point at a sibling package (`{"path": "../core"}`); its
    // files are outside the repo, and an import of one is an external, not a module.
    const outside = rel(file).startsWith("../");
    const skipped =
      outside ||
      !CODE.test(file) ||
      DECLARATION.test(file) ||
      file.split(path.sep).includes("node_modules");
    if (!skipped && !owners.has(file)) owners.set(file, parsed.options);
  }
}

const resolvers = new Map();
const files = [...owners.keys()]
  .sort((a, b) => (rel(a) < rel(b) ? -1 : rel(a) > rel(b) ? 1 : 0))
  .map((file) => {
    const options = owners.get(file);
    if (!resolvers.has(options)) resolvers.set(options, resolver(options));
    return scan(file, options, resolvers.get(options));
  });
process.stdout.write(`${JSON.stringify({ files })}\n`);
